homeassistant/components/zha/radio_manager.py

```diff
@@ -156,3 +212,171 @@ async def async_reset_adapter(self) -> None:
+        update is completed.
+        """
+        migration_data = HARDWARE_MIGRATION_SCHEMA(data)
+
+        name = migration_data["new_discovery_info"]["name"]
+        new_radio_type = ZhaRadioManager.parse_radio_type(
+            migration_data["new_discovery_info"]["radio_type"]
+        )
+        old_radio_type = ZhaRadioManager.parse_radio_type(
+            migration_data["old_discovery_info"]["radio_type"]
+        )
+
+        new_device_settings = new_radio_type.controller.SCHEMA_DEVICE(
+            migration_data["new_discovery_info"]["port"]
+        )
+        old_device_settings = old_radio_type.controller.SCHEMA_DEVICE(
+            migration_data["old_discovery_info"]["port"]
+        )
+
+        if (
+            self._config_entry.data[CONF_DEVICE][CONF_DEVICE_PATH]
+            != old_device_settings[CONF_DEVICE_PATH]
+        ):
+            # ZHA is using another radio, do nothing
+            return False
+
+        try:
+            await self._hass.config_entries.async_unload(self._config_entry.entry_id)
+        except config_entries.OperationNotAllowed:
+            # ZHA is not running
+            pass
+
+        # Temporarily connect to the old radio to read its settings
+        old_radio_mgr = ZhaRadioManager()
+        old_radio_mgr.hass = self._hass
+        old_radio_mgr.radio_type = old_radio_type
+        old_radio_mgr.device_path = old_device_settings[CONF_DEVICE_PATH]
+        old_radio_mgr.device_settings = old_device_settings
+        backup = await old_radio_mgr.async_load_network_settings(create_backup=True)
+
+        # Then configure the radio manager for the new radio to use the new settings
+        self._radio_mgr.chosen_backup = backup
+        self._radio_mgr.radio_type = new_radio_type
+        self._radio_mgr.device_path = new_device_settings[CONF_DEVICE_PATH]
+        self._radio_mgr.device_settings = new_device_settings
+        device_settings = self._radio_mgr.device_settings.copy()  # type: ignore[union-attr]
+
+        # Update the config entry settings
+        self._hass.config_entries.async_update_entry(
+            entry=self._config_entry,
+            data={
+                CONF_DEVICE: device_settings,
+                CONF_RADIO_TYPE: self._radio_mgr.radio_type.name,
+            },
+            options=self._config_entry.options,
+            title=name,
+        )
+        return True
+
+    async def async_finish_migration(self) -> None:
+        """Finish ZHA migration.
+
+        Throws an exception if the migration did not succeed.
+        """
+        # Restore the backup, permanently overwriting the device IEEE address
+        for retry in range(MIGRATION_RETRIES):
+            try:
+                if await self._radio_mgr.async_restore_backup_step_1():
+                    await self._radio_mgr.async_restore_backup_step_2(True)
+
+                break
+            except OSError as err:
```

Comment:
What the reason for failures commonly?

Reply:
The problem is that we don't know for sure when the addon is fully started